### PR TITLE
Hybridize

### PR DESCRIPTION
**Pull request by awikkerink**

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -27,14 +27,36 @@
...
+    "d2l-colors": "^3.0.1",
+    "d2l-offscreen": "^3.0.1",
+    "iron-iconset-svg": "polymerelements/iron-iconset-svg#^2.0.0",
+    "polymer": "Polymer/polymer#1.9 - 2"
   },
   "devDependencies": {
-    "d2l-demo-template": "^0.0.12",
-    "iron-test-helpers": "PolymerElements/iron-test-helpers#^1.4.1",
-    "web-component-tester": "^5.0.0"
+    "d2l-demo-template": "~1.0.1",
```
Comment (dlockhart) [inline bower.json]: Did you mean to replace the `^` with a `~`?

File: package.json [hunk 2]
```diff
@@ -6,7 +6,7 @@
     "build": "npm run build:scss && npm run build:iconsets",
     "build:scss": "node ./cli/create-scss-cli -o icons.scss images/**/*.*",
     "build:iconsets": "node ./cli/create-iconset-cli images/tier1 images/tier2 images/tier3 images/html-editor images/emoji",
-    "postinstall": "bower install",
+    "postinstall": "polymer install",
```
Comment (dlockhart) [inline package.json]: Why not `polymer install --variants` and not need to put that line in the `.travis.yml` file?

File: d2l-icon-button-shared-styles.html [hunk 3]
```diff
@@ -28,26 +29,36 @@
...
+			.d2l-icon-button:focus > d2l-icon {
 				@apply(--d2l-icon-button-icon-hover);
 			}
+
+			.d2l-icon-button > span {
+				@apply(--d2l-offscreen);
+			}
+			:host-context([dir="rtl"]) .d2l-icon-button > span {
+				@apply(--d2l-offscreen-rtl);
+			}
```
Comment (dlockhart) [inline d2l-icon-button-shared-styles.html]: This two rules could maybe be combined?

```
:host-context([dir="rtl"]) .d2l-icon-button > span,
:host(:dir(rtl)) .d2l-icon-button > span {
    @apply(--d2l-offscreen-rtl);
}
```

File: d2l-icon-button.html [hunk 4]
```diff
@@ -1,23 +1,20 @@
 <link rel="import" href="../polymer/polymer.html">
-<link rel="import" href="../d2l-offscreen/d2l-offscreen-shared-styles.html">
 <link rel="import" href="d2l-icon.html">
 <link rel="import" href="d2l-icon-button-shared-styles.html">
 <link rel="import" href="d2l-icon-button-behavior.html">
 
 <dom-module id="d2l-icon-button">
 	<template strip-whitespace>
```
Comment (dlockhart) [inline d2l-icon-button.html]: Now that the button is going inside, you may want to apply `display: inline-block` to the new `host` element.

File: d2l-icon-button-behavior.html [hunk 5]
```diff
@@ -5,6 +5,10 @@
 	/** @polymerBehavior */
 	var IconButtonBehavior = {
 
+		properties: {
+			disabled: Boolean
+		},
```
Comment (dlockhart) [inline d2l-icon-button-behavior.html]: Why was this needed?

File: bower.json [hunk 6]
```diff
@@ -27,14 +27,36 @@
     "package.json"
   ],
   "dependencies": {
-    "d2l-colors": "^2.3.0",
-    "d2l-offscreen": "^2.2.4",
-    "iron-iconset-svg": "polymerelements/iron-iconset-svg#^1.1.1",
-    "polymer": "^1.9.0"
+    "d2l-colors": "^3.0.1",
+    "d2l-offscreen": "^3.0.1",
```
Comment (dlockhart) [inline bower.json]: Should colors and offscreen be ranges as well? Just wondering how a `1.x` application would work, since it's going to be using the older version of colors. From [here](https://www.polymer-project.org/2.0/docs/devguide/hybrid-elements#dependency-variants):

>>> The default dependencies should use version ranges that include all versions supported by the component (typically, 1.9.1 or higher for Polymer itself).

File: d2l-icon-link.html [hunk 7]
```diff
@@ -1,23 +1,20 @@
 <link rel="import" href="../polymer/polymer.html">
-<link rel="import" href="../d2l-offscreen/d2l-offscreen-shared-styles.html">
 <link rel="import" href="d2l-icon.html">
 <link rel="import" href="d2l-icon-button-shared-styles.html">
 <link rel="import" href="d2l-icon-button-behavior.html">
```
Comment (dlockhart) [inline d2l-icon-link.html]: A bunch of problems here I think: a lot of the attributes that would have been on the old `<a>` element aren't getting moved over to the anchor -- `href`, `onclick`, etc. These will need to be transferred.

File: d2l-icon-button.html [hunk 8]
```diff
@@ -1,23 +1,20 @@
...
-		<style include="d2l-icon-button-shared-styles d2l-offscreen-shared-styles">
-			:host > span {
-				@apply(--d2l-offscreen);
-			}
+		<style include="d2l-icon-button-shared-styles">
 		</style>
-		<d2l-icon icon="[[icon]]"></d2l-icon>
-		<span><content></content></span>
+		<button class="d2l-icon-button" disabled={{disabled}}>
```
Comment (dlockhart) [inline d2l-icon-button.html]: Similar to my comment about `icon-link` -- all the attributes the consumer put on the `<button>` element before are now stuck on the `<d2l-icon-button>` and aren't getting transferred.

File: d2l-icon-button-shared-styles.html (hunk 3 above)
Comment (awikkerink) [inline d2l-icon-button-shared-styles.html]: In some browsers (Firefox I think), if one selector is invalid, the whole block is ignored.

File: d2l-icon-button-behavior.html (hunk 5 above)
Comment (awikkerink) [inline d2l-icon-button-behavior.html]: To pass the disabled attribute from d2l-icon-button to the button element

File: d2l-icon-link.html (hunk 7 above)
Comment (awikkerink) [inline d2l-icon-link.html]: `href` yes. The click event will still be handled.

File: bower.json [hunk 9]
```diff
@@ -27,14 +27,36 @@
...
+        "web-component-tester": "^5.0.0"
+      },
+      "resolutions": {
+        "webcomponentsjs": "^0.7",
+        "polymer": "Polymer/polymer#^1.9.1"
+      }
+    }
+  },
+  "resolutions": {
+    "webcomponentsjs": "0.7 - 1.0.0"
```
Comment (dlockhart) [inline bower.json]: Will all these ranges that don't have `^`s work with future minor/bug fix releases?

**Comment by awikkerink** [comment]
Maybe I'll change it to 0.7-1

File: d2l-icon-button.html [hunk 10]
```diff
@@ -1,28 +1,81 @@
...
+			autofocus={{autofocus}}
+			form={{form}}
+			formaction={{formaction}}
+			formenctype={{formenctype}}
+			formmethod={{formmethod}}
+			formnovalidate={{formnovalidate}}
+			formtarget={{formtarget}}
+			name={{name}}
+			type={{type}}
+			value={{value}}
```
Comment (dlockhart) [inline d2l-icon-button.html]: I think you can get away with just 1-way binding (i.e. `[[disabled]]`) since nothing is going to be interacting directly with the `<button>` element anymore to modify it. That'll be a performance gain.

File: d2l-icon-link.html [hunk 11]
```diff
@@ -1,28 +1,60 @@
...
-		<span><content></content></span>
+		<a class="d2l-icon-button"
+			title={{title}}
+			download={{download}}
+			href={{href}}
+			hreflang={{hreflang}}
+			rel={{rel}}
+			target={{target}}
+			type={{type}}
+		>
```
Comment (dlockhart) [inline d2l-icon-link.html]: Same comment here about using 1-way binding.
